### PR TITLE
feat: add soft margin loss

### PR DESCRIPTION
**Pull request by Bellman281**
### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

No existing issue — adds `SoftMarginLoss` for parity with `torch.nn.SoftMarginLoss`.

### Changes

Adds `SoftMarginLoss`, following the same `Config`-free `forward` /
`forward_no_reduction` pattern as `MseLoss`. Per-element loss is
`log(1 + exp(-target * logit))` = `softplus(-target * logit)`, with the target
tensor expected to hold values in `{-1, 1}`. No new dependencies; backward is
handled by autodiff.

### Testing

`cargo test -p burn-nn soft_margin` — known-answer test covering the un-reduced
output plus `Mean` and `Sum` reductions.

### CODE REVIEW

File: crates/burn-nn/src/loss/soft_margin.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
+use burn_core as burn;
+
+use crate::loss::reduction::Reduction;
+
+use burn::module::Module;
+use burn::tensor::Tensor;
+use burn::tensor::activation::softplus;
+
+/// Calculate the soft margin loss between the input logits and the target tensor.
```
Comment (Charles23R) [inline crates/burn-nn/src/loss/soft_margin.rs:9]: I would specify that it is specifically for a two-class classification task (hence the +1 / -1)

**Review by Charles23R — DISMISSED** [review]

**Review by Charles23R — APPROVED** [review]

File: crates/burn-nn/src/loss/soft_margin.rs (hunk 1 above)
Comment (Bellman281) [reply crates/burn-nn/src/loss/soft_margin.rs:9]: @Charles23R 
"Good point — updated the doc to specify this is a two-class (binary) classification criterion, which is why targets are ±1. Thanks!",

**Review by Bellman281 — COMMENTED** [review]
"Good point — updated the doc to specify this is a two-class (binary) classification criterion, which is why targets are ±1. Thanks!",
